pkg/driver/vol_statter.go

```diff
@@ -0,0 +1,169 @@
+}
+
+type VolStatterImpl struct {
+}
+
+func NewVolStatter() VolStatter {
+	return &VolStatterImpl{}
+}
+
+func (v VolStatterImpl) computeVolumeMetrics(volId, volPath string, refreshRate float64) (*volMetrics, error) {
+	if value, ok := v.retrieveFromCache(volId); ok {
+		if time.Since(value.timeStamp).Minutes() > refreshRate {
+			// Time to refresh volume stats
+			v.launchVolStatsRoutine(volId, volPath)
+		}
+		return value, nil
+	} else {
+		klog.V(4).Infof("Did not find volume metrics in cache for vol ID: %v , vol path: %v. Computing now!", volId, volPath)
+	}
+
+	v.launchVolStatsRoutine(volId, volPath)
+
+	// Return nil as kubelet might timeout waiting for volume stats
+	klog.Warningf("Volume metrics computation is underway for Vol ID: %v and metrics are not available yet.", volId)
+	return &volMetrics{
+		volPath:   volPath,
+		timeStamp: time.Now(),
+		volUsage: []*csi.VolumeUsage{
+			{
+				Unit: csi.VolumeUsage_UNKNOWN,
+			},
+		},
+	}, nil
+}
+
+func (v VolStatterImpl) retrieveFromCache(volId string) (*volMetrics, bool) {
+	if value, ok := volUsageCache[volId]; ok {
+		return value, true
+	} else {
+		return nil, false
+	}
+}
+
+func (v VolStatterImpl) removeFromCache(volId string) {
+	delete(volUsageCache, volId)
+}
+
+func (v VolStatterImpl) launchVolStatsRoutine(volId, volPath string) {
+	fsId, _, _, err := parseVolumeId(volId)
+	if err != nil {
+		klog.Errorf("Failed to launch Stat routine: Could not parse File System ID from volume Id - %s.", volId)
+		return
+	}
+
+	if _, ok := volStatterJobTracker[volId]; ok {
+		klog.V(5).Infof("Volume stats computation job is underway for volume Id : %v. Awaiting results", volId)
+	} else {
+		if ok := canStatFS(fsId); ok {
+			volStatterJobTracker[volId] = true
+			go v.computeDiskUsage(fsId, volId, volPath)
+		} else {
+			klog.V(5).Infof("Too many stat routines are running against FS : %s. Retry stat for volume Id: %s later", fsId, volId)
+		}
+	}
+}
+
+func (v VolStatterImpl) computeDiskUsage(fsId, volId, volPath string) {
+	waitTime := wait.Jitter(jitter, 2.0)
+	klog.V(5).Infof("Compute Volume Metrics invoked for Vol ID: %v, Sleeping for %v before execution", volId, waitTime)
+
+	//jittered execution
+	time.Sleep(waitTime)
+
+	used, err := fs.DiskUsage(volPath)
+	if err != nil {
+		klog.Errorf("Failed to compute volume usage on path %s: %v", volPath, err)
+		return
+	}
+
+	volUsed, ok := used.AsInt64()
+	if !ok {
+		klog.Errorf("Could not get volume used size: %v", err)
+		return
+	}
+
+	available, capacity, _, _, _, _, err := fs.FsInfo(volPath)
+	if err != nil {
+		klog.Errorf("Failed to fetch FsInfo on volume path %s: %v", volPath, err)
+		return
+	}
+
+	usage := []*csi.VolumeUsage{
+		{
+			Unit:      csi.VolumeUsage_BYTES,
+			Used:      volUsed,
+			Available: available,
+			Total:     capacity,
+		},
+	}
+
+	volMetrics := &volMetrics{
+		volPath:   volPath,
+		timeStamp: time.Now(),
+		volUsage:  usage}
+
+	volUsageCache[volId] = volMetrics
+	delete(volStatterJobTracker, volId)
```

Comment:
The map updates from this function are data races since the function is always called in a goroutine. The check in `launchVolStatsRoutine` to see if the `volId` is currently being processed is not sufficient - the check could be called at the same time as the delete here. I would suggest wrapping them in a sync mechanism. Removing the `time.Sleep(waitTime)` and stubbing out the external calls to the `fs` library will expose the data races in the test cases.

Reply:
I'd say if it's invoked infrequently, then the overhead of a lock would be negligible. What's the recovery path if this panics?

Reply:
I see what you mean. Didn't realize go panics on concurrent reads and writes. Alternative would be to either use `locks` or `sync.map`. I guess `locks` are better given how kubelet works?

Reply:
Correct, they are [not safe for concurrent use](https://golang.org/doc/faq#atomic_maps) and will panic if under certain conditions. I'm not overly familiar with `sync.map` but I believe it was created for optimization [under certain conditions](https://golang.org/pkg/sync/#Map) when regular locks are not meeting your performance needs.